src/v/cloud_storage/tests/cloud_storage_e2e_test.cc

```diff
@@ -583,3 +627,247 @@ FIXTURE_TEST(
+            } else {
+                auto abort_res = stm.abort_tx(pid, model::tx_seq(0), 1s).get();
+                BOOST_REQUIRE_EQUAL(abort_res, cluster::tx_errc::none);
+            }
+        }
+
+        // Roll the segment
+        auto log = dynamic_cast<storage::disk_log_impl*>(
+          partition->log().get());
+        log->flush().get();
+        log->force_roll(ss::default_priority_class()).get();
+
+        BOOST_REQUIRE(archiver->sync_for_tests().get());
+        archiver->upload_next_candidates().get();
+    }
+
+    BOOST_REQUIRE_EQUAL(
+      cloud_storage::upload_result::success,
+      archiver->upload_manifest("test").get());
+    archiver->flush_manifest_clean_offset().get();
+
+    vlog(
+      e2e_test_log.info,
+      "Test manifest size is {} bytes",
+      partition->archival_meta_stm()->manifest().segments_metadata_bytes());
+}
+
+ss::future<bool> check_consume_from_beginning_no_check(
+  kafka::client::transport client,
+  const model::topic& topic_name,
+  ss::gate& gate) {
+    try {
+        kafka_consume_transport consumer(std::move(client));
+        co_await consumer.start();
+        int iters = 0;
+        while (iters == 0 || !gate.is_closed()) {
+            auto holder = gate.hold();
+            auto kvs = co_await consumer.consume_from_partition(
+              topic_name, model::partition_id(0), model::offset(0));
+            vlog(e2e_test_log.debug, "Fetch returned {} records", kvs.size());
+            iters++;
+        }
+        co_return true;
+    } catch (...) {
+        vlog(
+          e2e_test_log.error,
+          "Consumer exception: {}",
+          std::current_exception());
+    }
+    co_return false;
+}
+
+FIXTURE_TEST(
+  test_consume_with_random_evictions, cloud_storage_manual_e2e_test) {
+    tests::remote_segment_generator gen(make_kafka_client().get(), *partition);
+    auto total_records
+      = gen.num_segments(100).batches_per_segment(10).produce().get();
+    BOOST_REQUIRE_GE(total_records, 1000);
+
+    // Compact the local log to GC to the collectible offset.
+    ss::abort_source as;
+    storage::housekeeping_config housekeeping_conf(
+      model::timestamp::min(),
+      1,
+      log->stm_manager()->max_collectible_offset(),
+      ss::default_priority_class(),
+      as);
+    partition->log()->housekeeping(housekeeping_conf).get();
+    // Wait for storage GC to remove local segments
+    tests::cooperative_spin_wait_with_timeout(30s, [this] {
+        auto log = dynamic_cast<storage::disk_log_impl*>(
+          partition->log().get());
+        return log->segments().size() == 1;
+    }).get();
+
+    ss::gate g;
+
+    std::vector<ss::future<bool>> checks;
+    checks.reserve(11);
+    for (int i = 0; i < 10; i++) {
+        checks.push_back(check_consume_from_beginning_rand(
+          std::move(make_kafka_client().get()), topic_name, g));
+    }
+    checks.emplace_back(evict_all(
+      app.cloud_storage_api.local(), app.shadow_index_cache.local(), g));
```

Comment:
If I'm reading this right we are only evicting once and hoping for this to interleave with the reads. Maybe we could periodically evict in a different fiber while the reads are ongoing?